### PR TITLE
remove azure hpp

### PR DESCRIPTION
**Pull request by vhvb1989**
Remove Azure.hpp

Make Azure/Core/Strings.hpp public

Move ToLower and InvariantIsEqual to Strings.hpp

Also removing `AZURE_UNREFERENCED_PARAMETER`  

fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/811

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/strings.hpp [hunk 1]
```diff
@@ -2,25 +2,16 @@
 // SPDX-License-Identifier: MIT
 
 /**
- * @file
- * @brief Common definitions.
+ * @brief Public functions for strings.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/strings.hpp:5]:
```suggestion
 * @brief Public utility functions for strings.
```

File: sdk/core/azure-core/inc/azure/core/strings.hpp [hunk 2]
```diff
@@ -2,25 +2,16 @@
...
-#pragma once
-
-#include "azure/core/internal/contract.hpp"
-
 #include <string>
 
-/**
- * @brief Used in implementations to mark an unreferenced function parameter.
- */
-#define AZURE_UNREFERENCED_PARAMETER(x) ((void)(x));
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/strings.hpp:18]: Why are we removing it?

File: sdk/core/azure-core/inc/azure/core/strings.hpp [hunk 3]
```diff
@@ -2,25 +2,16 @@
 // SPDX-License-Identifier: MIT
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/strings.hpp:2]: Add CL entries for these changes.

File: sdk/core/azure-core/src/http/curl/curl.cpp [hunk 4]
```diff
@@ -2,7 +2,7 @@
 // SPDX-License-Identifier: MIT
 
 #include "azure/core/http/curl/curl.hpp"
-#include "azure/core/azure.hpp"
+
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl.cpp:5]:
```suggestion
```

File: sdk/core/azure-core/src/http/transport_policy.cpp [hunk 5]
```diff
@@ -10,7 +13,10 @@ std::unique_ptr<RawResponse> TransportPolicy::Send(
     Request& request,
     NextHttpPolicy nextHttpPolicy) const
 {
-  AZURE_UNREFERENCED_PARAMETER(nextHttpPolicy);
+  if (!nextHttpPolicy.LastPolicy())
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/transport_policy.cpp:16]: This change doesn't seem relevant to the goal to "remove azure hpp".
Can we split this off into its own PR please?

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
Need CL update, why are we removing `AZURE_UNREFERENCED_PARAMETER` and maybe split off the `LastPolicy` change into its own separate PR.

File: sdk/storage/azure-storage-common/src/shared_key_policy.cpp [hunk 6]
```diff
@@ -16,20 +17,19 @@ namespace Azure { namespace Storage {
...
+                                   "Content-Language",
+                                   "Content-Length",
+                                   "Content-MD5",
+                                   "Content-Type",
+                                   "Date",
+                                   "If-Modified-Since",
+                                   "If-Match",
+                                   "If-None-Match",
+                                   "If-Unmodified-Since",
+                                   "Range"})
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/src/shared_key_policy.cpp:30]: why did you make these changes?

File: sdk/core/azure-core/src/http/transport_policy.cpp (hunk 5 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/src/http/transport_policy.cpp:16]: +1

File: sdk/core/azure-core/inc/azure/core/strings.hpp (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/strings.hpp:18]: there is only one use of it and I am replacing it to use `(void)parameter` instead as all other places.

File: sdk/core/azure-core/src/http/transport_policy.cpp (hunk 5 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/transport_policy.cpp:16]: it is somehow related. After removing AZURE_UNREF_PARAM(...) , we need to void the parameter.
But that's fine, let's have it apart :)

File: sdk/storage/azure-storage-common/src/shared_key_policy.cpp (hunk 6 above)
Comment (vhvb1989) [reply sdk/storage/azure-storage-common/src/shared_key_policy.cpp:30]: clang-formating

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/shared_key_policy.cpp:30]: this is weird, because I also clang-formatted it before.

File: sdk/core/azure-core/CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,9 @@
 ### Breaking Changes
 
 - `Azure::Core::Http::Url::AppendPath` now does not encode the input by default.
+- `azure.hpp` renamed to `strings.hpp`
+- `Azure::Core::Details::LocaleInvariantCaseInsensitiveEqual` moved to `Azure::Core::Strings::LocaleInvariantCaseInsensitiveEqual`
+- `Azure::Core::Details::ToLower` moved to `Azure::Core::Strings::ToLower`
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:10]: This isn't a breaking change. We are making a private method public. This should go under new features.

File: sdk/core/azure-core/CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,9 @@
 ### Breaking Changes
 
 - `Azure::Core::Http::Url::AppendPath` now does not encode the input by default.
+- `azure.hpp` renamed to `strings.hpp`
+- `Azure::Core::Details::LocaleInvariantCaseInsensitiveEqual` moved to `Azure::Core::Strings::LocaleInvariantCaseInsensitiveEqual`
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:9]: Same here, not a breaking change.

File: sdk/core/azure-core/CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,9 @@
 ### Breaking Changes
 
 - `Azure::Core::Http::Url::AppendPath` now does not encode the input by default.
+- `azure.hpp` renamed to `strings.hpp`
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:8]: I think this should be:
```suggestion
- Removed `azure.hpp`.
```

And under the features heading:
- Added strings.hpp with the \<fill this in\> tolower and other methods.

File: sdk/core/azure-core/inc/azure/core/strings.hpp [hunk 10]
```diff
@@ -2,25 +2,16 @@
...
 
 /**
- * @file
- * @brief Common definitions.
+ * @brief Public utility functions for strings.
+ *
  */
 
-#pragma once
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/strings.hpp:9]: Don't we need the `#pragma once`?

**Review by RickWinter — APPROVED** [review]

File: sdk/core/azure-core/CHANGELOG.md (hunk 9 above)
Comment (RickWinter) [reply sdk/core/azure-core/CHANGELOG.md:8]: I'd use what Victor has since the type is moving from azure.hpp to strings.hpp

File: sdk/core/azure-core/CHANGELOG.md (hunk 8 above)
Comment (RickWinter) [reply sdk/core/azure-core/CHANGELOG.md:9]: Ashon is correct, It's technically not a breaking change b/c it was in details namespace.

File: sdk/core/azure-core/CHANGELOG.md (hunk 7 above)
Comment (RickWinter) [reply sdk/core/azure-core/CHANGELOG.md:10]: Things in ::Details namespaces aren't private.  They are simply methods/details which have an implicit "do not use" contract.

Comment (RickWinter) [reply sdk/core/azure-core/CHANGELOG.md:10]: New feature makes sense to me as well

File: sdk/core/azure-core/CHANGELOG.md (hunk 9 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/CHANGELOG.md:8]: Which type? All we have are private methods in there.

File: sdk/core/azure-core/CHANGELOG.md [hunk 11]
```diff
@@ -5,6 +5,10 @@
 ### Breaking Changes
 
 - `Azure::Core::Http::Url::AppendPath` now does not encode the input by default.
+- Removed `azure.hpp`.
+
+### New Features
+- Added `strings.hpp` with `Azure::Core::Details::LocaleInvariantCaseInsensitiveEqual` and `Azure::Core::Details::ToLower`.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:11]: Not details.

```suggestion
- Added `strings.hpp` with `Azure::Core::Strings::LocaleInvariantCaseInsensitiveEqual` and `Azure::Core::Strings::ToLower`.
```

**Review by ahsonkhan — APPROVED** [review]
